doc/book/api.md

```diff
@@ -0,0 +1,194 @@
+
+```php
+class Request
+{
+    public function __construct(
+        $uri = null,
+        $method = null,
+        $body = 'php://memory',
+        array $headers = []
+    );
+
+    // See psr/http-message's RequestInterface for other methods
+}
+```
+
+Requests are immutable. Any methods that would change state -- those prefixed
+with `with` and `without` -- all return a new instance with the changes
+requested.
+
+## ServerRequest Message
+
+For server-side applications, `Zend\Diactoros\ServerRequest` implements
+`Psr\Http\Message\ServerRequestInterface`, which provides access to the elements
+of an HTTP request, as well as uniform access to the various elements of
+incoming data. The methods included are:
+
+```php
+class ServerRequest
+{
+    public function __construct(
+        array $serverParams = [],
+        array $fileParams = [],
+        $uri = null,
+        $method = null,
+        $body = 'php://input',
+        array $headers = []
+    );
+
+    // See psr/http-message's ServerRequestInterface for other methods.
+}
+```
+
+The `ServerRequest` is immutable. Any methods that would change state -- those
+prefixed with `with` and `without` -- all return a new instance with the changes
+requested. Server parameters are considered completely immutable, however, as
+they cannot be recalculated, and, rather, is a source for other values.
+
+## Response Message
+
+`Zend\Diactoros\Response` provides an implementation of
+`Psr\Http\Message\ResponseInterface`, an object to be used to aggregate response
+information for both HTTP clients and server-side applications, including
+headers and message body content. It includes the following:
+
+```php
+class Response
+{
+    public function __construct(
+        $body = 'php://memory',
+        $statusCode = 200,
+        array $headers = []
+    );
+
+    // See psr/http-message's ResponseInterface for other methods
+}
+```
+
+Like the `Request` and `ServerRequest`, responses are immutable. Any methods
+that would change state -- those prefixed with `with` and `without` -- all
+return a new instance with the changes requested.
+
+### ServerRequestFactory
+
+This static class can be used to marshal a `ServerRequest` instance from the PHP
+environment. The primary entry point is
+`Zend\Diactoros\ServerRequestFactory::fromGlobals(array $server, array $query,
+array $body, array $cookies, array $files)`. This method will create a new
+`ServerRequest` instance with the data provided. Examples of usage are:
+
+```php
+// Returns new ServerRequest instance, using values from superglobals:
+$request = ServerRequestFactory::fromGlobals();
+
+// or
+
+// Returns new ServerRequest instance, using values provided (in this
+// case, equivalent to the previous!)
+$request = RequestFactory::fromGlobals(
+  $_SERVER,
+  $_GET,
+  $_POST,
+  $_COOKIE,
+  $_FILES
+);
+```
+
+## URI
+
+`Zend\Diactoros\Uri` is an implementation of `Psr\Http\Message\UriInterface`,
```

Comment:
This was literally cut-and-paste from the old readme. Clearly, pushing docs
into their own directory and files prompts more thorough review! :-)
On May 26, 2015 6:36 PM, "Maks3w" <notifications@github.com> wrote:

> In doc/book/api.md
> <https://github.com/zendframework/zend-diactoros/pull/25#discussion_r31090738>
> :
>
> > +// or
> > +
> > +// Returns new ServerRequest instance, using values provided (in this
> > +// case, equivalent to the previous!)
> > +$request = RequestFactory::fromGlobals(
> > +  $_SERVER,
> > +  $_GET,
> > +  $_POST,
> > +  $_COOKIE,
> > +  $_FILES
> > +);
> > +```
> > +
> > +## URI
> > +
> > +`Zend\Diactoros\Uri` is an implementation of `Psr\Http\Message\UriInterface`,
>
> hyperlink to UriInterface
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/zendframework/zend-diactoros/pull/25/files#r31090738>.
>